### ISSUE TITLE
[bug] Some resources are not dropped when tauri app is closed

### ISSUE THREAD

**Issue by ehorrent** [tauri-apps/tauri#7358]
### Describe the bug

Hello !

I absolutely need to "cleanup" some resources when my app is closed. For that purpose, I implemented `Drop` trait on my managed state and discovered `drop` was never called.

Even worse, **any variable created in main function will never be destroyed** as it should. It seems any code after `run` call will never be reached (but in this situation I still expect my objects to be destroyed). 

I only did my tests on windows 11 and don't know if it can be reproduced on other platforms.
Perhaps I'm missing something, but if someone has a solution (or an explanation), I take it!

### Reproduction

1. Create an empty app using React / Vite / Typescript template
```powershell
yarn create tauri-app
cd tauri-app
yarn
```

2. Update main.rs
```rust
// Prevents additional console window on Windows in release, DO NOT REMOVE!!
#![cfg_attr(not(debug_assertions), windows_subsystem = "windows")]

use std::io::Write;

struct FakeState {
}

impl Drop for FakeState {
    fn drop(&mut self) {
        // Just an example, code here will never be executed
        let mut file = std::fs::File::create("c:\\temp\\drop.txt").unwrap();
        file.write_all(b"State dropped").unwrap();
    }
}

type SyncFakeState = std::sync::Mutex<FakeState>;

fn main() {
    let tauri_state = SyncFakeState::new(FakeState{});

    // Even local variables created in main will never be destroyed properly
    let other_state= SyncFakeState::new(FakeState{});

    tauri::Builder::default()
        .manage(tauri_state)
        .run(tauri::generate_context!())
        .expect("error while running tauri application");

   // Code here will never be reached
   panic!("App is closed");
}

```

3. Build the app and execute it
```powershell
yarn tauri build
cd ./src-tauri/target/release/
tauri-app.exe
```

4. Close the app, `FakeState` instances are not dropped

### Expected behavior

When the app is closed:
- managed state should be dropped
- any variable created in main should be dropped.

Solutions which could solve this issue:
- Tauri app exits gracefully and all variables are dropped **as expected**.
- or Tauri provides a way to catch a "close app" event.

### Platform and versions

```text
$ tauri info

[✔] Environment
    - OS: Windows 10.0.22621 X64
    ✔ WebView2: 114.0.1823.67
    ✔ MSVC: Visual Studio Community 2019
    ✔ rustc: 1.69.0 (84c898d65 2023-04-16)
    ✔ Cargo: 1.69.0 (6e9a83356 2023-04-12)
    ✔ rustup: 1.26.0 (5af9b9484 2023-04-05)
    ✔ Rust toolchain: stable-x86_64-pc-windows-msvc (default)
    - node: 16.14.0
    - yarn: 1.22.19
    - npm: 8.3.1

[-] Packages
    - tauri [RUST]: 1.4.1
    - tauri-build [RUST]: 1.4.0
    - wry [RUST]: 0.24.3
    - tao [RUST]: 0.16.2
    - @tauri-apps/api [NPM]: 1.4.0
    - @tauri-apps/cli [NPM]: 1.4.0

[-] App
    - build-type: bundle
    - CSP: unset
    - distDir: ../dist
    - devPath: http://localhost:1420/
    - framework: React
    - bundler: Vite
```

**Comment by FabianLars** [comment]
> It seems any code after run call will never be reached

Correct, if the app exits it will call [`process::exit`](https://doc.rust-lang.org/std/process/fn.exit.html) which doesn't run any Drop implementations or any of the code after the `.run()` line.

Currently you have 2 choices, either handle all the cleanup in tauri's ExitRequested and/or Exit events (this thingy here <https://github.com/tauri-apps/tauri/blob/b02fc90f450ff9e9d8a35ee55dc1beced4957869/examples/api/src-tauri/src/main.rs#L240-L244>) or use [`run_iteration`](https://docs.rs/tauri/latest/tauri/struct.App.html#method.run_iteration) instead which won't use `exit()` to close the program but also won't run tauri's internal cleanup (currently this entails cleaning up children spawned via tauri's Command api (does not include rust's std or tokio commands) and the system tray).

I _think_ we also played with the idea to just not call exit but just close the event loop but idk if we made any decisions. One thing that blocks it for me a bit is that then the restart command would behave differently from normal exits which may be even more unexpected (as opposed to all types of exits behaving the same)
